shell/server/pre-meteor.js

```diff
@@ -548,32 +549,70 @@ Meteor.startup(() => {
 
   WebApp.httpServer.removeAllListeners("request");
   WebApp.httpServer.on("request", (req, res) => {
-    try {
+    function nonMeteor() {
+      // Otherwise, dispatch to our own middleware proxy chain.
+      nonMeteorRequestHandler(req, res);
+      // Adjust timeouts on proxied requests to allow apps to long-poll if needed.
+      WebApp._timeoutAdjustmentRequestCallback(req, res);
+    }
+
+    Promise.resolve(undefined).then(() => {
       if (!req.headers.host) {
         res.writeHead(400, { "Content-Type": "text/plain" });
         res.end("Missing Host header");
         return;
       }
 
-      const hostname = req.headers.host.split(":")[0];
+      const hostname = req.headers.host.split(":")[0]; // strip port if it exists
       if (isSandstormShell(hostname)) {
+        if (req.url.startsWith("/_oauth/")) {
+          const parsedUrl = Url.parse(req.url, true);
+          if (parsedUrl.query && parsedUrl.query.state) {
+            const rawState = new Buffer(parsedUrl.query.state, "base64");
+            const state = JSON.parse(rawState.toString());
+            if (state.redirectUrl) {
+              const parsedRedirect = Url.parse(state.redirectUrl);
+              const redirectHostname = parsedRedirect.hostname;
+              return inMeteor(function () {
+                if (globalDb.hostIsStandalone(redirectHostname)) {
+                  res.writeHead(302, { "Location": parsedRedirect.protocol + "//" +
+                    parsedRedirect.host + req.url, });
+                  res.end();
+                  return;
+                } else {
+                  throw new Meteor.Error(400, "redirectUrl in OAuth was for an unknown host: " +
+                    state.redirectUrl);
+                }
+              });
+            }
+          }
+        }
         // If destined for the DDP host or the main host, pass on to Meteor
         for (let i = 0; i < meteorRequestListeners.length; i++) {
           meteorRequestListeners[i](req, res);
         }
       } else {
-        // Otherwise, dispatch to our own middleware proxy chain.
-        nonMeteorRequestHandler(req, res);
-        // Adjust timeouts on proxied requests to allow apps to long-poll if needed.
-        WebApp._timeoutAdjustmentRequestCallback(req, res);
+        return inMeteor(function () {
+          if (globalDb.hostIsStandalone(hostname)) {
```

Comment:
Can we do some sort of caching to avoid this DB query on every HTTP request?

Something like:
- If it's a wildcard host and the host ID is in the Proxy map, we know it's an app host.
- We can similarly maintain a map of known grain public IDs, and avoid any lookups if it's in there.
- Finally, we can also cache known standalone hosts, so that only the first request to a host hits the DB.